dovecot/save_net.cgi

```diff
@@ -12,20 +12,50 @@ $sslopt = &find("ssl_disable", $conf, 2) ? "ssl_disable" : "ssl";
 &save_directive($conf, $sslopt, $in{$sslopt} eq '' ? undef : $in{$sslopt});
 @listens = &find("imap_listen", $conf, 2) ?
 		("imap_listen", "pop3_listen", "imaps_listen", "pop3s_listen") :
-		("listen", "ssl_listen");
+		("listen");
 foreach $l (@listens) {
 	if ($in{$l."_mode"} == 0) {
 		$listen = undef;
 		}
 	elsif ($in{$l."_mode"} == 1) {
-		$listen = "[::]";
+		$listen = "*, ::";
 		}
 	elsif ($in{$l."_mode"} == 2) {
 		$listen = "*";
 		}
+	elsif ($in{$l."_mode"} == 4) {
+		$listen = "::";
+		}
 	elsif ($in{$l."_mode"} == 3) {
-		&check_ipaddress($in{$l}) || &error($text{'net_e'.$l});
-		$listen = $in{$l};
+		# Check each IP address
+		my @ips_list = split(/[\s,]+/, $in{$l});
+		my @ips_valid;
+		my $has_ip4_wildcard = grep { $_ eq "*" } @ips_list;
+		my $has_ip6_wildcard = grep { /^(\[::\]|::)$/ } @ips_list;
+		foreach my $ip (@ips_list) {
+			# Check for wildcards
+			if ($ip =~ /^(\*|::|\[::\])$/) {
+				push(@ips_valid, $ip);
+				next;
+				}
+			
+			# Validate IP address
+			my $is_ipv4 = &check_ipaddress($ip);
+			my $is_ipv6 = &check_ip6address($ip);
+			if (!$is_ipv4 && !$is_ipv6) {
+				&error(&text("net_ealisten", $ip));
+				}
+
+			# Add IP address to list
+			push(@ips_valid, $ip);
+
+			# Validate against wildcards
+			&error(&text("net_ealisten_invalid_mix", $ip, "*"))
+				if ($has_ip4_wildcard && &check_ipaddress($ip));
+			&error(&text("net_ealisten_invalid_mix", $ip, "::"))
+				if ($has_ip6_wildcard && &check_ip6address($ip));
```

Comment:
What case are these wildcard checks looking for?

Reply:
It's intended to prevent saving IPv4 or IPv6 addresses with same wildcard type, such as * and ::.

To be clear, it's not allowed to have `1.2.3.4, *` or `aaaa::bbbb::11, ::`, while it's okay to have `1.2.3.4, ::` or `aaaa::bbbb::11, *`.

Reply:
Ah ok makes sense